File: sdk/go/thrall/describe_test.go
package thrall

import (
	"reflect"
	"testing"
)

// TestOpNames proves the handler-map keys become a sorted op list, and an empty map yields nil so
// the describe field is omitted on the wire.
func TestOpNames(t *testing.T) {
	call := map[string]CallFn[int]{"value": nil, "get": nil}
	if got := opNames(call); !reflect.DeepEqual(got, []string{"get", "value"}) {
		t.Errorf("call ops = %v, want [get value] (sorted)", got)
	}
	cast := map[string]CastFn[int]{"reset": nil, "inc": nil}
	if got := opNames(cast); !reflect.DeepEqual(got, []string{"inc", "reset"}) {
		t.Errorf("cast ops = %v, want [inc reset] (sorted)", got)
	}
	if got := opNames(map[string]CastFn[int]{}); got != nil {
		t.Errorf("empty map ops = %v, want nil (omitted on the wire)", got)
	}
}

// TestFSMDescribe proves an FSM reports the union of its states' reaction ops - each dispatchable as
// a call or a cast - plus the reserved _state call op, all sorted, and passes the version through.
func TestFSMDescribe(t *testing.T) {
	def := FSM[int]{
		Version: "2.0.0",
		States: map[string]State[int]{
			"idle":    {On: map[string]Reaction[int]{"start": {}}},
			"running": {On: map[string]Reaction[int]{"stop": {}, "start": {}}}, // start repeats across states
		},
	}
	d := def.describe()
	if !reflect.DeepEqual(d.CallOps, []string{"_state", "start", "stop"}) {
		t.Errorf("call ops = %v, want [_state start stop]", d.CallOps)
	}
	if !reflect.DeepEqual(d.CastOps, []string{"start", "stop"}) {
		t.Errorf("cast ops = %v, want [start stop] (no reserved _state on the cast side)", d.CastOps)
	}
	if d.Version != "2.0.0" {
		t.Errorf("version = %q, want 2.0.0", d.Version)
	}
}

File: sdk/ts/src/describe.test.ts
import { test, expect } from "bun:test";
import { fsmDescribe, type FSMDef } from "./fsm";

// Mirrors the Go TestFSMDescribe: an FSM reports the union of its states' reaction ops - each
// dispatchable as a call or a cast - plus the reserved _state call op, all sorted, and the version.
test("fsmDescribe reports the union of reaction ops as call and cast, plus _state and version", () => {
  const def: FSMDef<number> = {
    name: "gate",
    initial: "idle",
    init: () => 0,
    version: "2.0.0",
    states: {
      idle: { on: { start: { fn: () => ({ data: 0 }) } } },
      running: { on: { stop: { fn: () => ({ data: 0 }) }, start: { fn: () => ({ data: 0 }) } } },
    },
  };
  const d = fsmDescribe(def);
  expect(d.call_ops).toEqual(["_state", "start", "stop"]);
  expect(d.cast_ops).toEqual(["start", "stop"]);
  expect(d.version).toBe("2.0.0");
});

// A machine with no reactions still exposes the reserved _state call op, and omits the empty cast set
// and the version - matching the Go omitempty wire form.
test("fsmDescribe on an empty machine has only _state and omits empties", () => {
  const def: FSMDef<number> = {
    name: "bare",
    initial: "s",
    init: () => 0,
    states: { s: { on: {} } },
  };
  const d = fsmDescribe(def);
  expect(d.call_ops).toEqual(["_state"]);
  expect(d.cast_ops).toBeUndefined();
  expect(d.version).toBeUndefined();
});
